types/mypy.ini

```diff
@@ -0,0 +1,78 @@
+[mypy]
+
+# strict mode enables all these if passed on the command line, but in
+# config files you have to specify them all individually.
+warn_unused_configs = True
+disallow_subclassing_any = True
+disallow_any_generics = True
+disallow_untyped_calls = True
+disallow_untyped_defs = True
+disallow_incomplete_defs = True
+check_untyped_defs = True
+disallow_untyped_decorators = True
+no_implicit_optional = True
+warn_redundant_casts = True
+warn_unused_ignores = True
+warn_return_any = True
+no_implicit_reexport = True
+
+
+strict_optional = True
+
+# --START-LAX-CONFIG--
+# Everything below this gets stripped out in order to generate a stricter version of this config.
+# see types/osh-parse.sh, typecheck-all function
+
+[mypy-core.alloc]
+strict_optional = False
+
+[mypy-core.main_loop]
+strict_optional = False
+
+[mypy-core.pyutil]
+strict_optional = False
+
+[mypy-core.ui]
+strict_optional = False
+
+[mypy-_devbuild.gen/syntax_asdl]
+strict_optional = False
+
+[mypy-frontend.match]
+strict_optional = False
+
+[mypy-frontend.parse_lib]
+strict_optional = False
+
+[mypy-frontend.tdop]
+strict_optional = False
+
+[mypy-oil_lang.expr_parse]
+strict_optional = False
+
+[mypy-oil_lang.expr_to_ast]
+strict_optional = False
+
+[mypy-osh.bool_parse]
+strict_optional = False
+
+[mypy-osh.braces]
+strict_optional = False
+
+[mypy-osh.cmd_parse]
+strict_optional = False
+
+[mypy-osh.glob_]
+strict_optional = False
+
+[mypy-osh.word_]
+strict_optional = False
+
+[mypy-osh.word_parse]
+strict_optional = False
+
+[mypy-pgen2.parse]
+strict_optional = False
+
+[mypy-pylib.os_path]
+strict_optional = False
```

Comment:
Hm I kind of feel like this isn't worth it... because we're going to add 100% type checking coverage in the near future, and then every time we'll have to edit this file.

That is, will every new file need to be entered here?

It seems like it's easier to pass --no-strict-optional once.  If I'm understanding the patch correctly.

It might be nice to turn on strict optional in the future, but it's much more important  to translate Oil to C++ first.

Reply:
Ah, ok.  I have heard of people going for a "ratchet effect" where you keep making things stricter.  That's what I was trying to do -- but no worries if it's overkill.  I guess the manifest file is serving a similar purpose, by enabling type checking one module at a time?

Anyway, main point is it's not necessary to enter every new file here.  You can specify modules as glob patterns - that's what I did originally, and I only needed a few.  Shall I switch back to that approach?

Reply:
@andychu maybe c1f70139087153c68692dcf28db9c48b6613ec82 is an improvement?  I believe this addresses your comment about ASDL being strict as well, below.  Note that syntax_asdl.py seems to need --no-strict-optional, in contrast to the rest of _devbuild/gen.

If you still think it's not worth the change after this, just say so, I won't be offended :)  Last thing I want to do is to take your attention away from the big stuff :)

Reply:
Yeah I'm not convinced we need a MyPy config right now.  The shell function is a hack, but it's a short one.  That's why I use shell -- to make problems go away :)

And generally I prefer flags over config files because it's less syntax to learn.  I'd rather use shell globs and a text file of files than globs in N different config syntaxes.

It's possible we might a config in the future, in which case we can dig this PR up again.  But I haven't needed to adjust MyPy flags in a long time.  I think we can complete the experiment with translating the parser without MyPy config changes, only code changes.

Reply:
Ok, no worries.  I think I disagree in one sense, but agree in another.

Disagree in the sense that it the mypy command line opts aren't as expressive as the config file.  You can specify just a few files to type check with particular options, but since mypy follows imports, you're pulling in a dependency graph, which will probably overlap with the dependency graph of other modules you want to type check with different options.  That's what the caused the `asdl/pretty.py` problem, I think.  It's getting checked with `strict_optional` set to both true and false in different situations, and so you either get an error about the None or an error about the unnecessary "ignore this None" comment.  

Using a config file to specify options on a module level lets you use the same options for `asdl.pretty` in both cases.  Of course, it would be nice if there were a way to get around this on the command line, or if mypy just wasn't so nitpicky about the unnecessary comment.

But I totally agree in the sense that it's a minor issue that won't hold up translating the parser.  And I'm sympathetic to the desire to avoid more syntaxes.

I'll close the MR, and get on with more useful stuff :)